src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+
+    def setattr_shopt(self, key, val):
+        """
+        set job attribute: shell option
+        """
+        self.setattr("system.shell.options." + key, val)
+
+    def dumps(self):
+        return json.dumps(self.jobspec)
+
+    @property
+    def resources(self):
+        return self.jobspec.get("resources", None)
+
+    @property
+    def tasks(self):
+        return self.jobspec.get("tasks", None)
+
+    @property
+    def attributes(self):
+        return self.jobspec.get("attributes", None)
+
+    @property
+    def version(self):
+        return self.jobspec.get("version", None)
+
+    def resource_walk(self):
+        """
+        Traverse the resources in the `resources` section of the jobspec.
+
+        Performs a depth-first, pre-order traversal.
+        """
+
+        def walk_helper(res_list):
+            for resource in res_list:
+                yield resource
+                children = resource.get("with", [])
+                # TODO: convert to `yield from` after dropping 2.7
+                for res in walk_helper(children):
+                    yield res
+
+        return walk_helper(self.resources)
+
+    def resource_walk_type(self, res_type):
+        """
+        Traverse the resources of type `res_type` in the `resources` section of the
+        jobspec.
+
+        Performs a depth-first, pre-order traversal.
+
+        :param res_type: resource type(s) to filter by
+        :type res_type: string or iterable of strings
+        """
+        if isinstance(res_type, six.string_types):
+            func = lambda x: x["type"] == res_type
+        elif isinstance(res_type, collectionsAbc.Iterable):
+            func = lambda x: x["type"] in res_type
+        else:
+            raise TypeError("res_type must be a string or an iterable")
+        return (x for x in self.resource_walk() if func(x))
+
+    def count_num_type(self, res_type):
+        def _count_helper(curr_res, res_type):
+            resource_count = curr_res["count"]
+            if curr_res["type"] == res_type:
+                return resource_count
+            num_in_children = sum(
+                _count_helper(r, res_type) for r in curr_res.get("with", [])
+            )
+            return resource_count * num_in_children
+
+        return sum(_count_helper(r, res_type) for r in self.jobspec["resources"])
+
+    def submit(self, flux_handle, flags=0):
+        return submit(flux_handle, self.dumps(), flags)
+
+
+class JobspecV1(Jobspec):
+    def __init__(self, resources, tasks, attributes=None, version=1):
+        """
+        Constructor for Version 1 of the Jobspec
+
+        :param version: included to allow for usage like JobspecV1(**jobspec)
+        """
+        if version != 1:
+            raise ValueError("version must be 1")
+        super(JobspecV1, self).__init__(
+            resources, tasks, version=version, attributes=attributes
+        )
+
+    @classmethod
+    def from_mini_syntax(
```

Comment:
Hmm... the name here doesn't feel quite right.  This builds a jobspecv1 from the components you would *get* from the mini syntax, rather than from either an argument list or a string of mini syntax.  I'm not sure if I'd prefer to change the name or have it take an argv list but probably one or ther other.

Reply:
I'd say leave the arg parsing in `mini` and change the name of the method to reflect its role as a simple jobspec constructor.  That was the point I was trying to make earlier.

Reply:
> I'd say leave the arg parsing in mini and change the name of the method to reflect its role as a simple jobspec constructor.

Sure.  Makes sense.  How do you feel about `from_command`? I'm thinking that since all the resource arguments are optional and `command` is the only required arg.

Reply:
Works for me.